js_modules/dagster-ui/packages/ui-core/src/live-data-provider/Factory.tsx

```diff
@@ -0,0 +1,54 @@
+import React from 'react';
+
+import {
+  LiveDataProvider,
+  LiveDataRefresh,
+  useLiveData,
+  useLiveDataSingle,
+} from './LiveDataProvider';
+import {LiveDataThreadID} from './LiveDataThread';
+import {LiveDataThreadManager} from './LiveDataThreadManager';
+
+export function liveDataFactory<T, R>(
+  useHooks: () => R,
+  queryKeys: (keys: string[], result: R) => Promise<Record<string, T>>,
+) {
+  const resultsFromUseHook: {current: R | undefined} = {current: undefined};
+  const manager = new LiveDataThreadManager((keys: string[]) => {
+    if (!resultsFromUseHook.current) {
+      throw new Error(
+        'Expected LiveDataProvider to have been in the DOM by the time queryKeys is called',
+      );
+    }
+    return queryKeys(keys, resultsFromUseHook.current);
```

Comment:
I wonder if `context` would be a better term for `result`? In the current impl, result is the Apollo Client, which might be confusing since result sounds like it'd be the live data